Sources/SourceKitLSP/Swift/DiagnosticReportManager.swift

```diff
@@ -0,0 +1,175 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LSPLogging
+import LanguageServerProtocol
+import SourceKitD
+import SwiftParserDiagnostics
+
+actor DiagnosticReportManager {
+  private let sourcekitd: SourceKitD
+  private let syntaxTreeManager: SyntaxTreeManager
+  private let documentManager: DocumentManager
+  private let clientHasDiagnosticsCodeDescriptionSupport: Bool
+
+  private nonisolated var keys: sourcekitd_keys { return sourcekitd.keys }
+  private nonisolated var requests: sourcekitd_requests { return sourcekitd.requests }
+
+  /// The cache that stores reports for snapshot id and buildSettings
+  ///
+  /// Conceptually, this is a dictionary. To prevent excessive memory usage we
+  /// only keep `cacheSize` entries within the array. Older entries are at the
+  /// end of the list, newer entries at the front.
+  private var reportCache:
+    [(
+      snapshotID: DocumentSnapshot.ID,
+      buildSettings: SwiftCompileCommand?,
+      report: RelatedFullDocumentDiagnosticReport
+    )] = []
+
+  /// The number of reports to keep
+  ///
+  /// - Note: This has been chosen without scientific measurements.
+  private let cacheSize = 5
+
+  init(
+    sourcekitd: SourceKitD,
+    syntaxTreeManager: SyntaxTreeManager,
+    documentManager: DocumentManager,
+    clientHasDiagnosticsCodeDescriptionSupport: Bool
+  ) {
+    self.sourcekitd = sourcekitd
+    self.syntaxTreeManager = syntaxTreeManager
+    self.documentManager = documentManager
+    self.clientHasDiagnosticsCodeDescriptionSupport = clientHasDiagnosticsCodeDescriptionSupport
+  }
+
+  func diagnosticReport(for snapshot: DocumentSnapshot, buildSettings: SwiftCompileCommand?, useCache: Bool = false)
+    async throws -> RelatedFullDocumentDiagnosticReport
```

Comment:
Nitpick: I prefer to break this on the arguments since it allows for easier extensions if we add more arguments in the future. Also, it looks nicer IMO.

```suggestion
  func diagnosticReport(
    for snapshot: DocumentSnapshot, 
    buildSettings: SwiftCompileCommand?, 
    useCache: Bool = false
  ) async throws -> RelatedFullDocumentDiagnosticReport
```